### ISSUE TITLE
Support LDAP sharing

### ISSUE THREAD

**Issue by roru69** [fcturner/passwords#175]
I use the "LDAP user and group backend" to authenticate against Owncloud.

When the local Admin or an "LDAP user tries to share a password with another "LDAP" user, the following message appears: "There are no users available you can share with."

When an "LDAP" user tries to share a password, only the local admin user is seen in the dialogbox. (sharing works).

Filesharing between "LDAP" users works just fine. The sharing api is enabled.

**Comment by fcturner** [comment]
Great, I'm learning something new here! Never go into LDAP before.
No idea how this works; aren't the users stored locally on your ownCloud database too?

**Comment by fcturner** [comment]
What does your oc_users table look like? Are the users stored there? Or is there a oc_ldap_users table or something like that? If it is, what are the column names?

**Comment by roru69** [comment]
I have attached some snapshots of the tables present. The user table only contains the "admin" user.


oc_group_user
![oc_group_user](https://cloud.githubusercontent.com/assets/9691381/14247721/7404ee8c-fa70-11e5-9e0e-2477b2bda35b.PNG)

oc_ldap_group_mapping
![oc_ldap_group_mapping](https://cloud.githubusercontent.com/assets/9691381/14247723/740637ec-fa70-11e5-9868-a10e59123185.PNG)

oc_ldap_group_members
![oc_ldap_group_members](https://cloud.githubusercontent.com/assets/9691381/14247724/74074218-fa70-11e5-87d0-73f20a411e63.PNG)

oc_ldap_user_mapping
![oc_ldap_user_mapping](https://cloud.githubusercontent.com/assets/9691381/14247725/740bf1fa-fa70-11e5-8c86-a1e4e5724527.PNG)

oc_users
![oc_users](https://cloud.githubusercontent.com/assets/9691381/14247722/74063e90-fa70-11e5-925c-c2c983e49a49.PNG)

**Comment by fcturner** [comment]
This is great! Are they logging in using `owncloud_name` from `oc_ldap_user_mapping`?

**Comment by fcturner** [comment]
Like Britney, Dylan, Edwin, and so on?

**Comment by roru69** [comment]
Yes, correct. that are their usernames in Active Directory (LDAP) and they are synchronised by the ldap plugin in Owncloud.

**Comment by roru69** [comment]
usernames are not case sensitive.

**Comment by fcturner** [comment]
So if I would use the UUID for each person, it should work?
```
Displayname		    uid			
------------------------------------
NULL -> admin		admin
britney				81EAB84E-1D....
dylan				2EEA34F0-44....
```

I will postpone this to 18.1, so basic sharing can be released for other users. Group sharing is something for 18.1 too.

Please keep in touch, I don't run LDAP, so I would really like you to test this if wrote the code :wink:

**Comment by roru69** [comment]
Sure thing. I will test the LDAP implementation for the Password app.. Thanks!!

**Comment by schannall** [comment]
If this will work I'll celebrate :)
Thank you for your hard work!

sC

**Comment by fcturner** [comment]
:wink:

**Comment by schannall** [comment]
Hi, any news on this? I'm still hoping to retire my old Password solution :)

sC

**Comment by fcturner** [comment]
Yes, working today on it! I want to release v19 before the end of June, and want it to support LDAP sharing, too.

**Comment by schannall** [comment]
This is great news. Thank you for your work! I'd like to help but testing will be the only thing I can do :(

sC

**Comment by roru69** [comment]
Same here, I can only test. Thanks for the effort you put in to this..

**Comment by fcturner** [comment]
Can someone using LDAP run this query?
```sql
SELECT 
	SUM(IF(appid = 'user_ldap'
			AND configkey = 'enabled'
			AND configvalue <> 'no',
		1,
		0)) AS ldap_enabled
FROM
	oc_appconfig
```

Output should be 
```
ldap_enabled
-------------
1
```

**Comment by fcturner** [comment]
(I need a test that reliably checks for LDAP usage)

**Comment by schannall** [comment]
It works as Intended:
`Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [c1owncloud]> SELECT
    ->     SUM(IF(appid = 'user_ldap'
    ->             AND configkey = 'enabled'
    ->             AND configvalue <> 'no',
    ->         1,
    ->         0)) AS ldap_enabled
    -> FROM
    ->     oc_appconfig
    -> ;
+--------------+
| ldap_enabled |
+--------------+
|            1 |
+--------------+
1 row in set (0.00 sec)
`

**Comment by fcturner** [comment]
Nice!

**Comment by fcturner** [comment]
First shot, please download the latest beta4:
https://github.com/fcturner/passwords/archive/ldap-support.zip
or
https://github.com/fcturner/passwords/archive/ldap-support.tar.gz

I would like to get info about screens or functions not working, javascript and PHP errors. I can't test this myself, so this is blind code. Thanks for testing!

**Comment by schannall** [comment]
Thanks a lot,
when enabeling: No app name specified
Renaming Folder to "passwords" works.

Nonl-LDAP User:
Opening the app with a non-ldap user works. There are currently 5 "passwords" (never had the app installed before) The Websites of those passwords corresponds to all LDAP Users* present in OC whose ID starts with a number.
Creating a Password works.
Sharing a password works only to LDAP Ids, that do not show up as "password" (E.g. starting with a letter).
Sharing passwords uses OC username, not loginname (which is bad for LDAP as it will normaly have username==ldapID)

*Wesite name == LDAP User ID 

so much for today :)

Jonas

**Comment by annunaki2k2** [comment]
Well done again @fcturner - this is something else that will be awesome once it's implemented. Keep up the great work!

I'll happily help with testing this out too (I use OpenLDAP, though I don't think that makes any difference).

For now, all I can report is that when I switched to the current ldap-support branch, I see a similar result as @schannall in that the LDAP users seem to be represented as password rows.

**Comment by fcturner** [comment]
@schannall Thank you for the great info. The screenshot helped a lot too!
Thanks for the compliment @annunaki2k2!

@ all, please try my newest version. And indeed, please rename the folder to `passwords`. I didn't update the version number, so it shouldn't take an ownCloud update procedure this time. Thanks for testing!

The users shouldn't pop up as websites anymore, and the names on the share dialog should be the LDAP names. I'm curious 😄 

https://github.com/fcturner/passwords/archive/ldap-support.zip
or
https://github.com/fcturner/passwords/archive/ldap-support.tar.gz

**Comment by annunaki2k2** [comment]
I've pulled this down and had a play. I can confirm that the users are no-longer appearing in the password list.

In addition, when I click to share a password, I can now see LDAP users listed, however I see only their UUIDs listed instead of their login names. Can we show their display name/login name instead? (I see the "displayname" is cached from LDAP in the MySQL oc_preferences table, so presumably there is just a getDisplayName sort of function to retrieve it?)

I currently can't test to see if the sharing has worked however, as I am unable to authenticate myself into the password app as an LDAP user. What happens is: I sucessfully log in to owncloud with an LDAP account & password, then switch to the password app and am prompted to re-enter the user password. Unfortunately, when I enter the correct password, I'm told "This password is invalid. Please try again.", and a (rather large) error is spewed to syslog. Here is what I think is just the relevant portion of the error:
`Aug  8 14:18:24 starbug ownCloud[24826]: {user_ldap} Exception: {"Exception":"Exception","Message":"No user available for the given login name.","Code":0,"Trace":"#0 \/var\/www\/annunaki2k2.co.uk\/htdocs\/owncloud\/apps\/user_ldap\/user_ldap.php(120): OCA\\user_ldap\\USER_LDAP->getLDAPUserByLoginName('604ca1df-db12-4...')\n
<<snip>>`

What you can see is that the function `getLDAPUserByLoginName` is being passed the UUID '604ca1df-db12-4...' (I assume it is the log output that has truncated the UUID, and not the app itself) rather than the actual login name, which in this instance should be "test".

Hope that is all clear and understood - if no please post questions and I'll do my best to answer. If you would like a full error message posting, just ask.

**Comment by roru69** [comment]
Installed the latest LDAP version of Passwords. No problem there.
Logged in as (local) Admin, activated the App and also no problems.

As (local)Admin I started the Password app and clicked on the share button.

![knipsel1](https://cloud.githubusercontent.com/assets/9691381/17699447/fdc0b05c-63c0-11e6-8674-ba3c8d147a01.PNG)

As you can see the UID's are shown and another thing I noticed is that when a user has a custom Avatar it is not shown in front of the UID.

I logged out the (Local) Admin and logged in as "AD" user.  In below sceenshot you can see that the password the (local) Admin shared is visible.

![knipsel2](https://cloud.githubusercontent.com/assets/9691381/17699643/2bf21b22-63c2-11e6-9779-4758ce788503.PNG)

 Sharing with the (local) Admin isn't possible because he is not listed,

![knipsel3](https://cloud.githubusercontent.com/assets/9691381/17699859/7ec7d1e2-63c3-11e6-82a6-1e6682666162.PNG)

Shared "geenstijl.nl" with another "AD" user. Logged out and logged in as the other user. Clicked on the shared password "share" button. (See screenshot).

![knipsel4](https://cloud.githubusercontent.com/assets/9691381/17700103/aea99e44-63c4-11e6-8cba-0b7fe2fcea4e.PNG)

As you can see the UID is shown and not the Username.

![knipsel6](https://cloud.githubusercontent.com/assets/9691381/17700196/04e5051e-63c5-11e6-8a52-e5c343f7aa3d.PNG)

Also no option to see the actual password. (bug or intended?).

When adding a new password and you forget to enter the URL and Sitename the "Error" dialogbox is shown behind the "New Password Entry" dialogbox.

![knipsel7](https://cloud.githubusercontent.com/assets/9691381/17700738/51a0d8ae-63c7-11e6-80fb-542dd7334e64.PNG)

LDAP sharing works, only a few small issues to solve..

Thanks!!

**Comment by schannall** [comment]
After this report I feel bad for my short one ;)!

sC

**Comment by fcturner** [comment]
@schannall Please don't! I'm very happy with all your replies!

@roru69 I'm extremely thrilled by your comments and thorough review!! I'm on holiday now, but I can fix these last things rather quickly at home I suppose.

**Comment by fcturner** [comment]
@roru69 @schannall  Please check latest master.
If fixed the alert box, share dialog and more, but can't check it since I don't run LDAP. Thanks for testing!

**Comment by fcturner** [comment]
Oh, avatars should be working too, please report back.

**Comment by fcturner** [comment]
@schannall I possibly fixed extra authentication with LDAP, please check that too :smile:

**Comment by annunaki2k2** [comment]
I've had a quick play, and still couldn't login with an LDAP user, however, I think you are very nearly there: I managed to successfully log in with an LDAP user by changing the pgrep_match to being case insensitive in service/authservice.php, an then replacing `\$username = OC::$server->getUserSession()->getUser()->getUID();` with `$username = \OC::$server->getUserSession()->getLoginName();`.

That was it - successful LDAP user login achieved! I have now successfully shared a test password from and LDAP user to another LDAP user, from and LDAP user to a local user, and from a local user to an LDAP user.

(You probably should change the if statement, and instead of checking for a GUID, do an `
\OC::$server->getUserSession()->getUser()-getBackendClassName() == "LDAP"` comparison - that seems somehow cleaner.)

That was the good news, however, the bad news is that the problem of the LDAP users displaying in the sharing popup as GUIDs rather than as their real names (display names), is still there.

I've tried to trace through the javascript, but unfortunately, I can't quite work out for the life of me what the source of the user data actually is. If you can point me in the right direction, I can debug it further, but for now, I'll have to leave that one to your good self.

**Comment by fcturner** [comment]
I didn't know 
```php 
\OC::$server->getUserSession()->getUser()->getBackendClassName() == "LDAP"
```
That's just perfect! In my opinion, ownCloud has horrible support for developers; I couldn't get this info from api.owncloud.org. Anyway, I'll fix it right away.

**Comment by fcturner** [comment]
Please try again 😄 
I added a View button for shared passwords (shared *to* you, not *by* you) and fixed the whole LDAP syntaxes in different files. I added an element to the Javascript page too, so I can reliably test for LDAP client side.

If the share dialog stills only shows GUIDs, you can do one of these options to debug:

**1. Edit source file**
Temporarily remove or comment [line 251 of /css/style.css](https://github.com/fcturner/passwords/blob/master/css/style.css#L251):
```css
#PasswordsTable,
#ShareUsersTable {
	display: none; /* delete or comment this one */
}
```

or
**2. Browser inspector**
find the element `<div id="ShareUsersTable">...</div>` and deselect `display: none;` in the right style element menu

![schermafbeelding 2016-09-06 om 10 01 48](https://cloud.githubusercontent.com/assets/7526559/18266123/5f3f2102-7419-11e6-900c-075a796de7fd.png)

**Comment by annunaki2k2** [comment]
Thanks for the extra bit of debugging there.

So, as you can see, both the uid and displayname are the same for the LDAP users:
![guids](https://cloud.githubusercontent.com/assets/3237224/18350131/5c052702-75cc-11e6-8d54-8434eedb7452.png)

Where (in the PHP) are you gathering that info from?

**Comment by fcturner** [comment]
Here: https://github.com/fcturner/passwords/blob/master/db/passwordmapper.php

I'm getting `owncloud_name` from `oc_ldap_user_mapping`, which should contain the right login name!
I'm forging it with the passwords and load the right username as 'website'. The variables should be renamed, but the system works.

**Comment by schannall** [comment]
@fcturner I'm not too sure about that. The username for Owncloud is (for example) 61FA0F90-624A-4BDC-[...].
Sharing uses the "Full name" when searching for the user, not the owncloud name..

Sorry for beeing late - I was on vacation and WLAN sucked :/

sC

**Comment by annunaki2k2** [comment]
@schannall , @fcturner : I'm been going through the user admin code to try and understand how that pulls down the list of full names (display names) out of LDAP for each user - that seems to be the only place I've found that show display names for LDAP users that I can find. You are right though: OwnCloud does not seem to store the display name of LDAP anywhere in the SQL db.

There is a getDisplayName() function for the OC_User class, and I assume that the admin pages probably iterate of the user IDs, instantiates each user, and then runs that function on each of them - that would take care of it, but it seems pretty heavy compared to the current approach, and how we best implement that in the password manager I don't quite know.

Perhaps a better approach is actually one similar to that of the file app, and have an search text box working with an AJAX call back to getDisplayName somewhere?

**Comment by schannall** [comment]
@annunaki2k2 Sorry not too much I can say. Maybe
https://github.com/owncloud/core/pull/2485
could help? It's quiet old but seems to have solved the Issue in the Fileshare back then.

sC

**Comment by fcturner** [comment]
@annunaki2k2 You're perfectly pointing towards a solution. This can be added to the PHP script that get the UUID's rather easily.
@schannall Good suggestion. Since 8.0.0, 
```php
\OC_User::getDisplayName($user)
```
is replaced by
```php
\OC::$server->getUserManager()->get($user)->getDisplayName()
```
, so I'll use that. The [API docs](https://doc.owncloud.org/api/classes/OCP.IUserManager.html#method_get) aren't very clear.

**Comment by fcturner** [comment]
Alright, please try the Pull Request :smile: 

Might break you installation...

**Comment by schannall** [comment]
I needed to update the installation after the installation of the password app:
```
Updating <passwords> ...
Updated <passwords> to 19-beta5
Drop old database tables
 Done
 28/28 [============================] 100%
Remove old (< 9.0) calendar/contact shares
 Done
 4/4 [============================] 100%
Fix permissions so avatars can be stored again
 Done
 2/2 [============================] 100%
Starting code integrity check...
Finished code integrity check
Update successful
```
Sharing a Password results in
`There are no users available you can share with`
But the log says the users are found: (owncloud.log)
```
Debug	user_ldap	Ready for a paged search	2016-09-13T09:43:17+02:00
Debug	user_ldap	initializing paged search for Filter objectClass=* base Array ( [0] => cn=xxxxxxxx,ou=users,ou=org,dc=xxxxxxx,dc=de ) attr Array ( [0] => displayName ) limit 500 offset 0	2016-09-13T09:43:17+02:00
Debug	user_ldap	readAttribute: cn=xxxxxxxx,ou=users,ou=org,dc=xxxxxxx,dc=de found	2016-09-13T09:43:17+02:00
Debug	user_ldap	Ready for a paged search
```
so much for today, thanks for you hard work!

sC

**Comment by fcturner** [comment]
Thanks for testing!
Anyone else too? My latest change in #257 shouldn't cause this. Did you try those fixes too?

**Comment by schannall** [comment]
Hi, I just cloned that repository.. There is only one OC user in our instance, the rest are all LDAP Users... I guess that it somehow won't display the LDAP names even if it finds them...
I will try again with a new local user.

sC

**Comment by annunaki2k2** [comment]
Seems I'm late to this party...

Well done @fcturner - I've pulled and switched to the `ldap-displaynames` branch, and can now see the LDAP users proper display names now!

The sharing pop-up now has display names correctly showing, however, at the moment , sadly I seem to have lost the ability to see/share with local users when logged in as an LDAP and vice-versa when logged on as a local user. I guess you now need to somehow merge the lists of users?

Now you know the trick, don't forget the "Shared by" & "Shared to" tool tips - they're still showing the GUID for LDAP usernames.

Well done again this great progress! :+1:

**Comment by schannall** [comment]
> The sharing pop-up now has display names correctly showing, however, at the moment , sadly I seem to have lost the ability to see/share with local users when logged in as an LDAP and vice-versa when logged on as a local user. I guess you now need to somehow merge the lists of users?

That explains why I did not find anyone - I was logged on as the local user.
Thanks for your help.

sC

**Comment by fcturner** [comment]
> Now you know the trick, don't forget the "Shared by" & "Shared to" tool tips - they're still showing the GUID for LDAP usernames.

Didn't expect that...

I'll fix it with combined with the local users.

**Comment by fcturner** [comment]
Please try again with this ldap branch 😄

**Comment by schannall** [comment]
As local User:
There are no users available you can share with. (Only one local User)
As LDAP User:
Fehler: Passwörter konnten nicht geladen werden. (Error: Passwords could not be read)

I don't know if this relates to this release.
Might this error occur because of my 2FA Authentication? Disableling 2FA did not help.
How can I clear the passwords for a User?

sC

**Comment by fcturner** [comment]
> As local User:
> There are no users available you can share with. (Only one local User)

Then please check your ownCloud settings. I tested the local users too with this branch and it works perfectly. Have you enabled sharing on the admin page? And allowed sharing with users in the same group?

I fixed the error! I missed a space. This part:
```php
FROM oc_ldap_user_mappingUNION ALL SELECT uid AS id
```
should be
```php
FROM oc_ldap_user_mapping UNION ALL SELECT uid AS id
```

Will fix it now.

**Comment by fcturner** [comment]
So @schannall your log resolved it all 😉

**Comment by fcturner** [comment]
Alright, please try again.

**Comment by fcturner** [comment]
> @schannall 
> How can I clear the passwords for a User?

You can push the button 'Move all passwords to trash bin' and permanently remove them there, or you can run this SQL script (since from your log I understand that you use MySQL):
```sql
DELETE FROM oc_passwords WHERE user_id = (username);
DELETE FROM oc_passwords_share WHERE shared_to = (username);
```

**Comment by schannall** [comment]
Thanks. Your new version works.
See the screenshot: Blue are LDAP Users, Red is a local user. One of the ldap Users is the one I'm logged on as.
![passwords](https://cloud.githubusercontent.com/assets/8982549/18502899/51d26cda-7a5a-11e6-9b1a-dab06c1795bd.PNG)

I'm not exactly sure where the 404 and 432 UIDs come from - they are not Users in OC at least ;)

The Local user can not share his password with anyone (remember: we only have one local user)

I'm looking for the two IDs and will report back.

**Comment by fcturner** [comment]
> The Local user can not share his password with anyone (remember: we only have one local user)

This is expected behaviour. I remove the active user from the list, as you shouldn't be able to share with yourself. Still need to implement that for LDAP, so thanks for noticing!

> Small error: Shared by 9EEE30FD-****.

How annoying! Thought I squeezed that bug.

I will first merge this into the master as it is now functional.

**Comment by schannall** [comment]
>This is expected behaviour. I remove the active user from the list, as you shouldn't be able to share with yourself. Still need to implement that for LDAP, so thanks for noticing!

Sure, but the LDAP Users should appear nevertheless :)

**Comment by fcturner** [comment]
Are you in the same group? Are you allowed to share with LDAP users as a local users according to admin share settings?

I will release v19 now, and these small fixed will be included in v19.1.
Thanks everybody!

**Comment by schannall** [comment]
Hi, somehow it does not work. I can share the passwords to other people but they won't see them. When I select the users in the Share dialog and save that the "Badge" will show the correct amount of shared Users (LDAP and Local). If I open the Share dialog again the LDAP User will not be checked (and won't have the Password). The Local User does recive the password.
I hope you understand what I mean.

sC
